### DISCUSSION TITLE
Relational patterns and boxing

### DISCUSSION THREAD

**Discussion by uladz-zubrycki** [dotnet/csharplang#5981] [Q&A] [answered]
I have just faced somewhat unexpected behavior in regard to relational patterns and boxing, would be happy to know if it's correct and why it is this way.

Result is `"less"` for boxed `integer`
```
object value = -1;
var result = value switch
{
	< 0 => "less",
	_ => null
};
```

And it's `null` for boxed `decimal` (same for `double`)
```
object value = -1m;
var result = value switch
{
	< 0 => "less",
	_ => null
};
```

But it's `less` again if literal in the pattern is of proper type (`0m` for `decimal`) 
```
object value = -1m;
var result = value switch
{
	< 0m => "less",
	_ => null
};
```

Or if `value` isn't boxed, i.e it's `decimal`
```
decimal value = -1m;
var result = value switch
{
	< 0 => "less",
	_ => null
};
```

I guess it's related to the fact that types of the boxed value and pattern are compared before the values are, but all this is quite confusing and inconsistent, considering that `< 0` works just fine for an non boxed `decimal` with implicit cast.

```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.101
 Commit:    ef49f6213a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.101\

Host (useful for support):
  Version: 6.0.1
  Commit:  3a25a7f1cc

.NET SDKs installed:
  3.1.100 [C:\Program Files\dotnet\sdk]
  3.1.417 [C:\Program Files\dotnet\sdk]
  6.0.100 [C:\Program Files\dotnet\sdk]
  6.0.101 [C:\Program Files\dotnet\sdk]
```

**Comment by svick** [comment, marked as answer]
This behaves [as specified](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/patterns3#relational-patterns):

> When the input is a type for which a suitable built-in binary relational operator is defined that is applicable with the input as its left operand and the given constant as its right operand, the evaluation of that operator is taken as the meaning of the relational pattern. Otherwise we convert the input to the type of the expression using an explicit nullable or unboxing conversion. It is a compile-time error if no such conversion exists. The pattern is considered not to match if the conversion fails.
